subprojects/diagnostics/src/main/java/org/gradle/api/tasks/diagnostics/internal/configurations/model/ConfigurationReportModelFactory.java

```diff
@@ -0,0 +1,212 @@
+    }
+
+    private ReportCapability convertDefaultCapability(Project project) {
+        return new ReportCapability(Objects.toString(project.getGroup()), project.getName(), Objects.toString(project.getVersion()), true);
+    }
+
+    /**
+     * This class can examine a project to deterimne which {@link Attribute}s in its schema have {@link org.gradle.api.internal.attributes.CompatibilityRule CompatibilityRule}s or
+     * {@link org.gradle.api.internal.attributes.DisambiguationRule DisambiguationRule}s defined.
+     */
+    private final class ConfigurationRuleScanner {
+        private final AttributesSchema attributesSchema;
+
+        public ConfigurationRuleScanner(Project project) {
+            attributesSchema = project.getDependencies().getAttributesSchema();
+        }
+
+        public List<ReportAttribute> getAttributesWithCompatibilityRules() {
+            return attributesSchema.getAttributes().stream()
+                .filter(this::hasCompatibilityRules)
+                .map(ConfigurationReportModelFactory.this::convertUncontainedAttribute)
+                .sorted(Comparator.comparing(ReportAttribute::getName))
+                .collect(Collectors.toList());
+        }
+
+        public List<ReportAttribute> getAttributesWithDisambiguationRules() {
+            return attributesSchema.getAttributes().stream()
+                .filter(this::hasDisambiguationRules)
+                .map(ConfigurationReportModelFactory.this::convertUncontainedAttribute)
+                .sorted(Comparator.comparing(ReportAttribute::getName))
+                .collect(Collectors.toList());
+        }
+
+        private boolean hasCompatibilityRules(Attribute<?> attribute) {
+            final AttributeMatchingStrategy<?> matchingStrategy = attributesSchema.getMatchingStrategy(attribute);
+            final DefaultCompatibilityRuleChain<?> ruleChain = (DefaultCompatibilityRuleChain<?>) matchingStrategy.getCompatibilityRules();
+            return ruleChain.doesSomething();
+        }
+
+        private boolean hasDisambiguationRules(Attribute<?> attribute) {
+            final AttributeMatchingStrategy<?> matchingStrategy = attributesSchema.getMatchingStrategy(attribute);
+            final DefaultDisambiguationRuleChain<?> ruleChain = (DefaultDisambiguationRuleChain<?>) matchingStrategy.getDisambiguationRules();
+            return ruleChain.doesSomething();
+        }
+    }
+
+    /**
+     * Recursively calling {@link Map#computeIfAbsent(Object, Function)} to insert multiple keys could pose issues, depending on the implementation of the underlying map.
+     *
+     * This method exists to externalize that usage and avoid the potential for these sort of issues.
+     */
+    private static <K, V> V computeIfAbsentExternal(Map<K, V> map, K key, Function<? super K, ? extends V> mappingFunction) {
+        if (map.containsKey(key)) {
+            return map.get(key);
```

Comment:
Personally I think that get/`null` check/return is a better idiom here (one less map access), but it's not a huge deal.
```suggestion
        V existing = map.get(key);
        if (existing != null) {
            return existing;
```

Reply:
I strongly disagree - I think when you have the option, code written to avoid making explicit null checks gains worthwhile clarity and limits cognitive load, even if it is slightly slower.  But this stuff isn't performance critical.